edgeless_orc/src/orchestrator.rs

```diff
@@ -725,6 +977,143 @@ impl Orchestrator {
+                    // functions/resources has disappeared and it was not
+                    // possible to fix the situation immediately).
+                    for (origin_ext_fid, instance) in active_instances.iter() {
+                        match instance {
+                            ActiveInstance::Function(start_req, _reqs, instances) => {
+                                let num_disconnected = instances.iter().filter(|x| to_be_disconnected.contains(&x.node_id)).count();
+                                assert!(num_disconnected <= instances.len());
+                                if instances.is_empty() || num_disconnected > 0 {
+                                    to_be_repatched.push(origin_ext_fid.clone());
+                                    if instances.is_empty() || num_disconnected == instances.len() {
+                                        // If all the function instances
+                                        // disappared, then we must enforce the
+                                        // creation of (at least) a new
+                                        // function instance.
+                                        fun_to_be_created.insert(origin_ext_fid.clone(), start_req.clone());
+                                    } else {
+                                        // Otherwise, we just remove the
+                                        // disappeared function instances and
+                                        // let the others still alive handle
+                                        // the logical function.
+                                        active_instances_to_be_updated.push(*origin_ext_fid);
+                                    }
+                                }
+                            }
+                            ActiveInstance::Resource(start_req, instance) => {
+                                if instance.is_none() || to_be_disconnected.contains(&instance.node_id) {
+                                    to_be_repatched.push(origin_ext_fid.clone());
+                                    res_to_be_created.insert(origin_ext_fid.clone(), start_req.clone());
+                                }
+                            }
+                        }
+                    }
+
+                    // Also schedule to repatch all the functions that
+                    // depend on the functions/resources modified.
+                    for (origin_ext_fid, output_mapping) in active_patches.iter() {
+                        for (_output, target_ext_fid) in output_mapping.iter() {
+                            if active_instances_to_be_updated.contains(target_ext_fid)
+                                || fun_to_be_created.contains_key(target_ext_fid)
+                                || res_to_be_created.contains_key(target_ext_fid)
+                            {
+                                to_be_repatched.push(origin_ext_fid.clone());
```

Comment:
Does this also deal with internal links?
The repatch seems to be depend on the external set.

Reply:
It is designed to patch everything as expected.

The gold information on what to patch (= the edges in the function dependency graph) is in `active_patches`.